cpp/src/link_prediction/similarity_impl.cuh

```diff
@@ -162,5 +163,302 @@ rmm::device_uvector<weight_t> similarity(
+  } else {
+    sources.resize(graph_view.local_vertex_partition_range_size(), handle.get_stream());
+    thrust::sequence(handle.get_thrust_policy(),
+                     sources.begin(),
+                     sources.end(),
+                     graph_view.local_vertex_partition_range_first());
+  }
+
+  if (topk) {
+    //  We can reduce memory footprint by doing work in batches and
+    //  computing/updating topk with each batch
+    rmm::device_uvector<vertex_t> top_v1(0, handle.get_stream());
+    rmm::device_uvector<vertex_t> top_v2(0, handle.get_stream());
+    rmm::device_uvector<weight_t> top_score(0, handle.get_stream());
+
+    top_v1.reserve(*topk, handle.get_stream());
+    top_v2.reserve(*topk, handle.get_stream());
+    top_score.reserve(*topk, handle.get_stream());
+
+    //   FIXME: Think about what this should be
+    edge_t const MAX_PAIRS{2 << 20};
+
+    rmm::device_uvector<edge_t> degrees = graph_view.compute_out_degrees(handle);
+    rmm::device_uvector<edge_t> two_hop_degrees(degrees.size(), handle.get_stream());
+
+    // Let's compute the maximum size of the 2-hop neighborhood of each vertex
+    // FIXME: If sources is specified, this could be done on a subset of the vertices
+    //
+    edge_dst_property_t<GraphViewType, edge_t> edge_dst_degrees(handle, graph_view);
+    update_edge_dst_property(handle, graph_view, degrees.begin(), edge_dst_degrees);
+
+    per_v_transform_reduce_incoming_e(
+      handle,
+      graph_view,
+      edge_src_dummy_property_t{}.view(),
+      edge_dst_degrees.view(),
+      edge_dummy_property_t{}.view(),
+      [] __device__(vertex_t, vertex_t, auto, auto dst_degree, auto) { return dst_degree; },
+      edge_t{0},
+      reduce_op::plus<edge_t>{},
+      two_hop_degrees.begin());
+
+    if (source_vertices) {
+      rmm::device_uvector<edge_t> gathered_degrees(sources.size(), handle.get_stream());
+
+      thrust::gather(
+        handle.get_thrust_policy(),
+        thrust::make_transform_iterator(
+          sources.begin(),
+          cugraph::detail::shift_left_t<vertex_t>{graph_view.local_vertex_partition_range_first()}),
+        thrust::make_transform_iterator(
+          sources.end(),
+          cugraph::detail::shift_left_t<vertex_t>{graph_view.local_vertex_partition_range_first()}),
+        two_hop_degrees.begin(),
+        gathered_degrees.begin());
+
+      two_hop_degrees = std::move(gathered_degrees);
+    }
+
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        two_hop_degrees.begin(),
+                        two_hop_degrees.end(),
+                        sources.begin(),
+                        thrust::greater<edge_t>{});
+
+    thrust::inclusive_scan(handle.get_thrust_policy(),
+                           two_hop_degrees.begin(),
+                           two_hop_degrees.end(),
+                           two_hop_degrees.begin());
+
+    size_t current_pos{0};
+    size_t next_pos{0};
+
+    while (true) {
```

Comment:
We may use https://github.com/rapidsai/cugraph/blob/branch-24.04/cpp/include/cugraph/utilities/misc_utils.cuh#L40

I think vertex_t and edge_t in (template <typename vertex_t, typename edge_t>) should better be renamed, but we can get all the chunk boundaries at once with this function.